src/controllers/v1/attachment/__tests__/index.test.ts

```diff
@@ -0,0 +1,304 @@
+import { expect } from 'chai'
+import sinon from 'sinon'
+import mockdate from 'mockdate'
+
+import { attachment } from '../index.js'
+import { Env } from '../../../../env.js'
+import Database from '../../../../lib/db/index.js'
+import { octetExample, jsonExample } from './fixtures.js'
+import Ipfs from '../../../../lib/ipfs.js'
+import { BadRequest, InternalServerError, NotFound } from '../../../../lib/error-handler/index.js'
+import { AttachmentRow } from '../../../../lib/db/types.js'
+
+mockdate.set('2020-10-10')
+
+describe('v1/attachment', () => {
+  let response: any
+  let controller: attachment
+  const database: Database = new Database()
+  const ipfs: Ipfs = new Ipfs(new Env())
+
+  const stubs = {
+    getFile: sinon
+      .stub(ipfs, 'getFile' as any)
+      .callsFake(async () => ({ blob: { arrayBuffer: () => '0'.repeat(1024) }, filename: 'a' })),
+    addFile: sinon.stub(ipfs, 'addFile' as any).callsFake(async () => 'QmXVStDC6kTpVHY1shgBQmyA4SuSrYnNRnHSak5iB6Eehn'),
+    get: sinon.stub(database, 'get').callsFake(async () => [octetExample]),
+    update: sinon.stub(database, 'update' as any).callsFake(async (_, updates: any) => [
+      {
+        ...octetExample,
+        ...updates,
+      },
+    ]),
+    insert: sinon
+      .stub(database, 'insert' as any)
+      .callsFake((_, data: any) => [{ ...data, created_at: new Date(), id: 'attachment-insert-test' }]),
+  }
+
+  before(() => {
+    controller = new attachment(database, ipfs)
+  })
+
+  after(() => {
+    stubs.get.reset()
+    stubs.getFile.reset()
+    stubs.insert.reset()
+    stubs.get.reset()
+    stubs.update.reset()
+  })
+
+  describe('create() - POST /', () => {
+    describe('if payload does not contain body or file', () => {
+      beforeEach(async () => {
+        response = await controller.create({} as any).catch((err) => err)
+      })
+
+      it('throws BadRequest error', () => {
+        expect(response).to.be.instanceOf(BadRequest)
+        expect(response.code).to.equal(400)
+        expect(response.message).to.equal('nothing to upload')
+      })
+
+      it('does not call ipfs', () => {
+        expect(stubs.addFile.callCount).to.equal(0)
+      })
+
+      it('does not update existing attachment', () => {
+        expect(stubs.update.callCount).to.equal(0)
+      })
+
+      it('and does not call database insert method', () => {
+        expect(stubs.insert.callCount).to.equal(0)
+      })
+    })
+
+    describe('if database insert method fails', () => {
+      beforeEach(async () => {
+        stubs.insert.throwsException(new InternalServerError('this.db.insert() failure'))
+
+        response = await controller.create({ body: { data: 'attachment-test' } } as any).catch((err) => err)
+      })
+
+      // QUESTION shall we rely on off-chain and do not assert here?
+      // reset back to top level stubbing
```

Comment:
Don't know .. that being said, should we leave this two lines here in the code ?

Reply:
Well, since they are test files I do not pay much attention. The above line have been removed